File: Arrays/Kadanes_Algorithm/Code.py
# Brute-Force Approach
def max_subarray_sum_bruteforce(nums):
    max_sum = float('-inf')
    n = len(nums)

    for i in range(n):
        for j in range(i, n):
            total = sum(nums[i:j+1])
            max_sum = max(max_sum, total)

    return max_sum

# Improved Brute-Force Approach
def max_subarray_sum_improved_bruteforce(nums):
    max_sum = float('-inf')
    n = len(nums)

    for i in range(n):
        total = 0
        for j in range(i, n):
            total += nums[j]
            max_sum = max(max_sum, total)

    return max_sum

# Divide and Conquer Approach
def max_subarray_sum_divide_conquer(nums):
    def max_crossing_sum(nums, left, mid, right):
        left_sum = float('-inf')
        total = 0
        for i in range(mid, left - 1, -1):
            total += nums[i]
            left_sum = max(left_sum, total)

        right_sum = float('-inf')
        total = 0
        for i in range(mid + 1, right + 1):
            total += nums[i]
            right_sum = max(right_sum, total)

        return left_sum + right_sum

    def max_subarray_sum_helper(nums, left, right):
        if left == right:
            return nums[left]

        mid = (left + right) // 2

        left_sum = max_subarray_sum_helper(nums, left, mid)
        right_sum = max_subarray_sum_helper(nums, mid + 1, right)
        crossing_sum = max_crossing_sum(nums, left, mid, right)

        return max(left_sum, right_sum, crossing_sum)

    return max_subarray_sum_helper(nums, 0, len(nums) - 1)

# Kadane's Algorithm
def max_subarray_sum_kadane(nums):
    max_so_far = float('-inf')
    max_ending_here = 0

    for num in nums:
        max_ending_here = max(num, max_ending_here + num)
        max_so_far = max(max_so_far, max_ending_here)

    return max_so_far

if __name__ == "__main__":
    arr = [-2, 1, -3, 4, -1, 2, 1, -5, 4]
    print("Brute-Force Approach:", max_subarray_sum_bruteforce(arr))
    print("Improved Brute-Force Approach:", max_subarray_sum_improved_bruteforce(arr))
    print("Divide and Conquer Approach:", max_subarray_sum_divide_conquer(arr))
    print("Kadane's Algorithm:", max_subarray_sum_kadane(arr))

File: Arrays/Kadanes_Algorithm/Code.cpp
#include <iostream>
#include <vector>

using namespace std;

// Brute-Force Approach
int maxSubArraySumBruteForce(vector<int>& nums) {
    int maxSum = INT_MIN;
    int n = nums.size();

    for (int i = 0; i < n; i++) {
        for (int j = i; j < n; j++) {
            int sum = 0;
            for (int k = i; k <= j; k++) {
                sum += nums[k];
            }
            maxSum = max(maxSum, sum);
        }
    }

    return maxSum;
}

// Improved Brute-Force Approach
int maxSubArraySumImprovedBruteForce(vector<int>& nums) {
    int maxSum = INT_MIN;
    int n = nums.size();

    for (int i = 0; i < n; i++) {
        int sum = 0;
        for (int j = i; j < n; j++) {
            sum += nums[j];
            maxSum = max(maxSum, sum);
        }
    }

    return maxSum;
}

// Divide and Conquer Approach
int maxCrossingSum(vector<int>& nums, int left, int mid, int right) {
    int sum = 0;
    int leftSum = INT_MIN;

    for (int i = mid; i >= left; i--) {
        sum += nums[i];
        leftSum = max(leftSum, sum);
    }

    sum = 0;
    int rightSum = INT_MIN;

    for (int i = mid + 1; i <= right; i++) {
        sum += nums[i];
        rightSum = max(rightSum, sum);
    }

    return leftSum + rightSum;
}

int maxSubArraySumDivideConquer(vector<int>& nums, int left, int right) {
    if (left == right) {
        return nums[left];
    }

    int mid = (left + right) / 2;

    int leftSum = maxSubArraySumDivideConquer(nums, left, mid);
    int rightSum = maxSubArraySumDivideConquer(nums, mid + 1, right);
    int crossingSum = maxCrossingSum(nums, left, mid, right);

    return max({leftSum, rightSum, crossingSum});
}

// Kadane's Algorithm
int maxSubArraySumKadane(vector<int>& nums) {
    int max_so_far = INT_MIN;
    int max_ending_here = 0;

    for (int num : nums) {
        max_ending_here = max(num, max_ending_here + num);
        max_so_far = max(max_so_far, max_ending_here);
    }

    return max_so_far;
}

int main() {
    vector<int> arr = { -2, 1, -3, 4, -1, 2, 1, -5, 4 };
    cout << "Brute-Force Approach: " << maxSubArraySumBruteForce(arr) << endl;
    cout << "Improved Brute-Force Approach: " << maxSubArraySumImprovedBruteForce(arr) << endl;
    cout << "Divide and Conquer Approach: " << maxSubArraySumDivideConquer(arr, 0, arr.size() - 1) << endl;
    cout << "Kadane's Algorithm: " << maxSubArraySumKadane(arr) << endl;

    return 0;
}
